File: Java/cp/Remove_Outermost_parentheses.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Stacks {
    public String removeOuterParentheses(String S) {
        if (S.length() == 0) return "";
        Stack<Character> stack = new Stack<>();
        int count = 0;
        
        for (int i = S.length() - 1; i > 0; i--)
            stack.push(S.charAt(i));
            
        String ans = "";
        while (!stack.isEmpty()){
            if (stack.peek() == '('){
                if (count >= 0)
                    ans+= stack.peek();
                count++;
            }
            else{
                if (count != 0)
                    ans += stack.peek();
                count --;
            }
            stack.pop();
        }
        return ans;
    }
}

/*
TIME COMPLEXITY: 0(N) where n is size of string
SPACE COMPLEXITY:0(N) where n is size of string
TEST CASE

INPUT
"(()())(())"

OUTPUT
"()()()"

INPUT
"(()())(())(()(()))"

OUTPUT
"()()()()(())"

*/

File: C/other/remove_outermost_parentheses.c
//algorithm for removing remove outermost parantheses in O(n) time complexity and O(1) space complexity

#include <stdio.h>

char * remove_outer_parantheses(char * S)
{
    int count = 0;
    int stack = 0;
    char *ans = S;
    char c; 
    while (c = *S++)
    {
        if (c == ')') 
        {
            --stack;
        }
        if (stack > 0) 
        {
            ans[count++] = c;
        }
        if (c == '(') 
        {
            ++stack;
        }
            
    }
    ans[count] = '\0';
    return ans;
}

int main()
{   //taking input from the user
    char s[1000];
    printf(" Enter the string\n ");
    scanf(" %s ",s);
    //printing result
    printf("The final string is : %s\n ",remove_outer_parantheses(s));
    return 0;
}

/*
Sample Input 1:
Input: "(()())(())"
Output: The final string is : ()()()
  
Sample Input 2:
Input: "(())()"
Output: The final string is : ()
*/